README.md
# webcollab
This is a git repository tracking my local git repository for WebCollab development.

The origin git repository can be found at http://sourceforge.net/p/webcollab/code/ci/master/tree/


webcollab/CHANGELOG
WebCollab Git
*** 29 October 2020

- Bugfix: Correct corrupted characters in Swedish language file (Thanks to Jan Roos)

- update language codes in translation files

- update translation README

WebCollab Git
*** 8 September 2020

- update calendar_show to include projects and parent tasks

- add tooltips to calendar_show, contact_menubox and task_navigate

- Bugfix: box_shorten() not splitting long words correctly

WebCollab 3.50 - Nihoniho
*** 2 September 2019

- release

WebCollab Git
*** 5 September 2019

- update French translation (Thanks to: martinbrait)

- add development debugging to config.php

WebCollab Git
*** 18 August 2019

- add token validation to archive

- add token validation to contacts

- refactor task_edit to handle archive menu

- refactor contact_submit

- refactor includes/time

WebCollab Git
*** 14 August 2019

- change to random_bytes() for session and token generation

- Bugfix:  Incorrect function call in smtp_auth (no error caused)

- Add etag and better headers to rss

WebCollab Git
*** 4 August 2019

- minimum PHP version to 7.0.0

- cast delta_deadline as int in task_clone_submit.php

WebCollab Git
*** 11 June 2019

- update French translation (Thanks to: martinbrait)

WebCollab Git
*** 28 April 2019

- change column names in database tables to avoid reserved words in MySQL 8

- update Esperanto translation (Thanks to: Stella Lindblom)

- Bugfix: PHP 7.2 gives an error for using arrays created as strings in icalendar_todo.php

WebCollab Git
*** 19 May 2018

- update email address

WebCollab Git
*** 6 May 2018

- change column 'admin' to 'user_admin' (admin is a reserved word in MySQL 8) (Thanks to: Sam Dodd)

WebCollab 3.46 - Rotorua
*** 2 September 2017

- release

WebCollab Git
*** 13 August 2017

- minor HTML improvements to suit HTML5

- provide code for HTML5 option in screen.php

- allow for setting smtp TLS port number in setup

WebCollab Git
*** 25 March 2017

- add mb_check_encoding() use before using regex

- improve setting UTF-8 encoding

WebCollab Git
*** 5 February 2017

- Bugfix: Links in email require login due to incorrect $_COOKIE check (Thanks to Paul Knyazev)

- add random_bytes() for PHP 7

- update minimum PHP to 5.6

WebCollab 3.45 - Akitio
*** 16 July 2016

- release

WebCollab Git
*** 27 April 2016

- Bugfix: Clone leaves all projects and tasks open and editable by usergroups (Thanks to: hguttzeit)

WebCollab 3.44 - Sweet Pea
*** 17 December 2015

- release

- based on WebCollab 3.42 with config version fixed.

WebCollab Git
*** 15 December 2015

- change to password_hash() for password functions

- automatically upgrade older md5 password hashes to blowfish crypt on login

- remove WORK_FACTOR constant for blowfish crypt.  Use PHP defaults.

- update minimum PHP to 5.5

WebCollab Git
*** 18 September 2015

- remove db_fetch_all() (Reverse 21 Febuary 2015 edit)

- Bugfix: CloseCursor() added to task_navigate.php to prevent 'ATTR_EMULATE_PREPARES' errors.

- add filter_var() for integer validation in common.php

- add filter_var() for email validation in admin_config_submit and user_submit

WebCollab 3.43 - Harrison 130
*** 19 June 2015

- release

WebCollab Git
*** 27 April 2015

- Bugfix:  Syntax error in code causes error in PHP 5.6+

WebCollab 3.42 - Cobb River Revisited
*** 19 February 2015

- release

WebCollab Git
*** 21 February 2015

- Bugfix:  Incorrect database call in task_submit_clone.php (Thanks to: bshah) (Bug #98)

- Bugfix:  Change database calls to db_fetch_all().  Prevents 'ATTR_EMULATE_PREPARES' errors.

WebCollab 3.41 - Cobb River
*** 2 June 2014

- release

WebCollab Git
*** 7 February 2015

- use better Blowfish algorithm

- enable openssl_random_pseudo_bytes() for all OS (Windows was fixed in PHP 5.3.4)

WebCollab Git
*** 2 November 2014

- Bugfix: Syntax errors in calendar date (Credit: Igor Falcomata)

WebCollab Git
*** 19 October 2014

- improve mysql_pdo connection code

- extend token timeout to 30 minutes by default

- improve token key generation

WebCollab Git
*** 12 October 2014

- improved salt generation using mt_rand() and base64_encode()

- use openssl_random_pseudo_bytes() for session key generation where available

- improve sesson key generation with Mersenne Twister fallback

WebCollab Git
*** 9 September 2014

- update minimum PHP to 5.4.0 (PHP 5.3.0 is at end of life)

- remove magic_quotes checks (not required in PHP 5.4.0+)

- default to blowfish password encryption

- use 'default_charset' option to set UTF-8 with PHP 5.6.0+

WebCollab Git
*** 3 September 2014

- bugfix: SQL constraint violation on login in index.php (Patch #37) (Credit: Luke Bigum)

WebCollab Git
*** 5 July 2014

- Updated Bulgarian translation (Credit: Svetlozar Ivanov )

WebCollab 3.40 - Glenorchy
*** 2 June 2014

- release

WebCollab Git
*** 26 February 2014

- Updated Italian translation (Credit: Giuseppe Dantini )

WebCollab Git
*** 19 January 2014

- Bug #97  - Fix for foreign cookies overwriting POST & GET by removing $_REQUEST superglobals. (Reported by: Favern)

- improve formatting to Forum Search.  Add to CSS code.

WebCollab Git
*** 3 January 2014

- time.php: change from preg function to mb_substr for better execution speed. Cache escaped date results.

WebCollab Git
*** 3 November 2013

- improve and simplify conformance to RFC 2045 of quoted printable encoding in emails

- simplify encoded headers in email using iconv_mime_encode()

- minor improvements to email headers

WebCollab 3.31 - Baldy
*** 18 October 2013

- bugfix release for HTTP Response Splitting Vulnerablity (CVE-2013-2652)

WebCollab Git
*** 17 October 2013

- Italian translation partially updated (Credit: Igor Falcomata )

- bugfix:  password_hash() is a reserved function in PHP 5.5 (Credit: Midov )

WebCollab Git
*** 28 July 2013

- change to storing uploaded files with hashed id

- add Blowfish password encryption for PHP > 5.3.7

WebCollab Git
*** 21 May 2013

- change md5() hashes to sha1() hashes


webcollab/index.php
<?php
/*
  $Id: index.php 2288 2009-08-22 08:50:00Z andrewsimpson $

  (c) 2002 - 2019 Andrew Simpson <andrewnz.simpson at gmail.com>

  WebCollab
  ---------------------------------------

  This program is free software; you can redistribute it and/or modify it under the
  terms of the GNU General Public License as published by the Free Software Foundation;
  either version 2 of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
  PARTICULAR PURPOSE. See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with this
  program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.

   Function:
  ---------

  Secure the login

*/

require_once('path.php' );
require_once(BASE.'path_config.php' );
require_once(BASE_CONFIG.'config.php' );

include_once(BASE.'lang/lang.php' );
include_once(BASE.'includes/common.php' );
include_once(BASE.'includes/screen.php' );

//error condition
function secure_error($error='Login error', $redirect_time=0 ) {

  global $lang;

  $content = "<div style=\"text-align : center\">".$error."</div>";
  create_top($lang['login'], 1, 0, 0, $redirect_time );
  new_box($lang['error'], $content, 'boxdata-small', 'head-small' );

  if($redirect_time != 0) {
    $content = "<div style=\"text-align : center\"><a href=\"".BASE_URL."index.php\">".$lang['login_now']."</a></div>\n";
    new_box(sprintf($lang['redirect_sprt'], $redirect_time ), $content, 'boxdata-small', 'head-small' );
  }

  create_bottom();
  die;
}

//enable login
function enable_login($userid, $username, $ip='0.0.0.0', $taskid ) {

  //create session key
  try {
    //generate 40 characters hex
    $session_key = bin2hex(random_bytes(20 ) );
  } 
  //if no suitable random number sources are available in the operating system, random_bytes() will throw an exception 
  catch(Exception $e ) {
    //use Mersenne Twister algorithm (random number) to give up to 128 bits, then one-way hash to give session key
    $session_key = sha1(mt_rand().mt_rand().mt_rand().mt_rand() );
  }

  //remove the old login information
  $q = db_prepare('DELETE FROM '.PRE.'logins WHERE user_id=?' );
  @db_execute($q, array($userid ) );
  $q = db_prepare('DELETE FROM '.PRE.'login_attempt WHERE last_attempt < (now()-INTERVAL '.db_delim('20 MINUTE' ).') OR login_name=?' );
  @db_execute($q, array($username ) );
  @db_query('DELETE FROM '.PRE.'tokens WHERE lastaccess < (now()-INTERVAL '.db_delim(TOKEN_TIMEOUT.' MINUTE' ).')' );

  //log the user in
  $q = db_prepare('INSERT INTO '.PRE.'logins(user_id, session_key, ip, lastaccess ) VALUES (?, ?, ?, now() )' );
  @db_execute($q, array($userid, $session_key, $ip ) );
  
  //try and set a session cookie (if the browser will let us)
  $url = parse_url(BASE_URL );
  //use HTTP only to reduce XSS attacks (only in PHP 5.2.0+ )
  setcookie('webcollab_session', $session_key, 0, $url['path'], $url['host'], false, true );
  //(No need to record an error here if unsuccessful: the code will revert to URI session keys)

  //relocate the user to the main screen
  //(we use both URI session key and cookies initially - in case cookies don't work)
  if($taskid == 0 ) {
    header('Location: '.BASE_URL.'main.php?x='.$session_key );
  }
  else {
    header('Location: '.BASE_URL.'tasks.php?x='.$session_key.'&action=show&taskid='.$taskid );
  }
  die;
  return;
}

//perform login query
function login_query($username ) {

  //construct login query
  if(! ($q = db_prepare('SELECT id FROM '.PRE.'users WHERE user_name=? AND deleted=\'f\'', 0 ) ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  //database query
  if( ! @db_execute($q, array($username ), 0 ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  //no such user-password combination
  if( ! ($userid = @db_result($q, 0, 0) ) ) {
    secure_error('Access denied to unknown user \''.$username.'\'' );
  }
  else {
    return $userid;
  }
  return false;
}

//record recent login failures
function record_fail($username, $ip ) {

 global $lang;

  //record this login attempt
  $q = db_prepare('INSERT INTO '.PRE.'login_attempt(login_name, ip, last_attempt ) VALUES (?, ?, now() )' );
  db_execute($q, array($username, $ip ) );

  //wait 2 seconds then record an error
  sleep (2);
  secure_error($lang['no_login'], 15 );
  die;
}

//limit number of login attempts
function check_lockout($username ) {

  //count the number of recent failed login attempts
  if(! ($q = db_prepare('SELECT COUNT(*) FROM '.PRE.'login_attempt WHERE login_name=?
			      AND last_attempt > (now()-INTERVAL '.db_delim('10 MINUTE').') LIMIT 6', 0 ) ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  if( ! @db_execute($q, array($username ), 0 ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  $count_attempts = db_result($q, 0, 0 );

  //protect against password guessing attacks
  if($count_attempts > 4 ) {
    secure_error("Exceeded allowable number of login attempts.<br /><br />Account locked for 10 minutes." );
    die;
  }

  return true;
}

function password_upgrade($userid, $password ) {

  $hash = password_hash($password, PASSWORD_DEFAULT );

  //check for valid hash
  if($hash === false ) return false;

  //update the password
  db_begin();

  $q = db_prepare('UPDATE '.PRE.'users SET user_password=? WHERE id=?' );
  db_execute($q, array($hash, $userid ) );

  db_commit();

  return true;
}

//
// MAIN LOGIN
//

//check and set taskid & nologin if required
if(isset($_GET['taskid']) && @safe_integer($_GET['taskid']) ) {
  $taskid  = $_GET['taskid'];
}
elseif(isset($_POST['taskid']) && @safe_integer($_POST['taskid']) ) {   
  $taskid = $_POST['taskid'];
}
else {
  $taskid = 0;
}

$nologin = (isset($_GET['nologin']) ) ? 1 : 0;

//secure variables
$content = '';
$q = '';
$row = '';
$hash = 'xxxx';
$salt = '';
$username = '0';
$password = '0';
$session_key = '';

//log ip address
if( ! ($ip = $_SERVER['REMOTE_ADDR'] ) ) {
  secure_error('Unable to determine ip address');
}

// 1. Password login authentication
if(isset($_POST['username']) && isset($_POST['password']) && strlen($_POST['username']) > 0 && strlen($_POST['password']) > 0 && ACTIVE_DIRECTORY != 'Y' ) {

  include_once(BASE.'database/database.php');

  $username = safe_data($_POST['username'] );

  //check for account locked
  check_lockout($username );

  //construct login query for username / password
  if(! ($q = db_prepare('SELECT id, user_password FROM '.PRE.'users WHERE user_name=? AND deleted=\'f\'', 0 ) ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  //database query
  if( ! @db_execute($q, array($username), 0 ) ) {
    secure_error('Unable to connect to database.  Please try again later.' );
  }

  //if user-password combination exists
  if($row = @db_fetch_array($q, 0 ) ) {

    //bcrypt encryption or SHA256 + salt (deprecated - used WebCollab 3.30 - 3.31 )
    if(strlen($row['user_password'] ) > 50 ){
      //verify password
      if(password_verify($_POST['password'], $row['user_password'] ) ) {
        //check need for rehash (work factor changed or older SHA256 )
        if(password_needs_rehash($row['user_password'], PASSWORD_DEFAULT ) ) {
          password_upgrade($row['id'], validate($_POST['password'] ) );
        }
        //valid password -> continue
        enable_login($row['id'], $username, $ip, $taskid );
      }
    }

    //fallback to older md5 encryption (deprecated - used WebCollab 1.01 - 3.30 )
    if(strlen($row['user_password'] ) < 35 ) {
      //verify password
      if($row['user_password'] === md5($_POST['password'] ) ) {
        //upgrade password now...
        password_upgrade($row['id'], validate($_POST['password'] ) );
        //valid password -> continue
        enable_login($row['id'], $username, $ip, $taskid );
      }
    }
  }

  //no such user-password combination
  record_fail($username, $ip);
}

// 2. Web authorisation
if(WEB_AUTH === 'Y' && isset($_SERVER['REMOTE_USER']) && (strlen($_SERVER['REMOTE_USER']) > 0 ) ) {

  include_once 'database/database.php';

  $username = safe_data($_SERVER['REMOTE_USER'] );

  if($userid = login_query($username ) ) {
    enable_login($userid, $username, $ip, $taskid );
  }
}

// 3. ACTIVE DIRECTORY login
if(ACTIVE_DIRECTORY == 'Y' && isset($_POST['username']) && isset($_POST['password']) && strlen($_POST['username']) > 0 && strlen($_POST['password']) > 0 ) {

  include_once(BASE.'database/database.php');

  $username = safe_data($_POST['username']);
  $password = safe_data($_POST['password'] );

  //check for account locked
  check_lockout($username );

  if(! $adconn = ldap_connect($AD_HOST, AD_PORT ) ) {
    secure_error('ACTIVE_DIRECTORY: Connection not successful.' );
  }

  ldap_set_option($adconn, LDAP_OPT_PROTOCOL_VERSION, 3 );

  if( ! $ldap_bind = ldap_bind($adconn, $username, $password ) ) {
    secure_error($lang['no_login'], 15 );
  }

  ldap_close($adconn );

  if($userid = login_query($username ) ) {
    enable_login($userid, $username, $ip, $taskid );
  }

  //record failure
  record_fail($username, $ip );

}

// 4. Continuation of session
if(isset($_COOKIE['webcollab_session'] ) && preg_match('/^[a-f\d]{40}$/i', $_COOKIE['webcollab_session'] ) && (! $nologin ) ) {
  //allow for continuation of session if a valid cookie is already set
  // if 'nologin' is set we have just been rejected by security.php

  include_once 'database/database.php';

  //check if session is valid and within time limits
  $q = db_prepare('SELECT COUNT(*) FROM '.PRE.'logins
                          WHERE session_key=?
                          AND lastaccess > (now()-INTERVAL '.db_delim(round(SESSION_TIMEOUT).' HOUR').')' );
  db_execute($q, array(safe_data($_COOKIE['webcollab_session']) ) );

  if(db_result($q, 0, 0 ) == 1 ) {

    //relocate to main screen, and let security.php do further checking on session validity
    if($taskid == 0 ) {
      header('Location: '.BASE_URL.'main.php' );
    }
    else {
      header('Location: '.BASE_URL.'tasks.php?action=show&taskid='.$taskid );
    }
    die;
  }
}

//
// LOGIN SCREEN
//

//create login screen
create_top($lang['login_screen'], 1, 'login', 0 );

$content = "<div style=\"text-align:center\">\n";

if(SITE_IMG != '' && SITE_IMG != 'webcollab.png' ) {
  $content .=  "<p><img src=\"images/".SITE_IMG."\" alt=\"Site logo\" /></p>\n";
}
else {
  $content .=  "<p><img src=\"images/webcollab.png\" alt=\"WebCollab logo\" width=\"322\" height=\"102\" /></p>\n";
}

$content .= "<p>".$lang['please_login'].":</p>\n".
	    "<form method=\"post\" action=\"index.php\">\n".
	    "<fieldset><input type=\"hidden\" name=\"taskid\" value=\"".$taskid."\" /></fieldset>\n".
	    "<table style=\"margin-left:auto; margin-right:auto;\">\n".
	    "<tr style=\"text-align: left\" ><td>".$lang['login'].": </td><td><input id=\"username\" class=\"size\" type=\"text\" name=\"username\" value=\"\" />".
	    "<script type=\"text/javascript\">document.getElementById('username').focus();</script></td></tr>\n".
	    "<tr style=\"text-align: left\" ><td>".$lang['password'].": </td><td><input type=\"password\" class=\"size\" name=\"password\" value=\"\" /></td></tr>\n".
	    "</table>\n".
	    "<p style=\"padding-top: 20px; padding-bottom: 20px\"><input type=\"submit\" value=\"".$lang['login_action']."\" /></p>\n".
	    "</form>\n";

switch(DATABASE_TYPE ) {

  case 'postgresql_pdo':
    $content .= "<p><a href=\"http://www.postgres.org\"><img src=\"images/postgresql-power.png\" width=\"80\" height=\"15\" alt=\"Powered by postgresql\" /></a></p>\n";
    break;

  case 'mysql_pdo':
    $content .= "<p><a href=\"http://www.mysql.com\"><img src=\"images/poweredbymysql-125.png\" width=\"125\" height=\"42\" alt=\"Powered by MySQL\" /></a></p>\n";
    break;

  default:
    $content .= "<p><a href=\"http://www.php.net\"> <img src=\"images/php-power.png\" width=\"88\" height=\"31\" alt=\"Powered by PHP\" /></a></p>\n";
    break;
}

$content .= "</div>\n";

//set box options
new_box($lang['login'], $content, 'boxdata-small', 'head-small' );

create_bottom();

?>
